# 01. Data Loading and Exploration | تحميل واستكشاف البيانات

## 📚 Learning Objectives

By completing this notebook, you will:
- Understand the key concepts of this topic
- Apply the topic using Python code examples
- Practice with small, realistic datasets or scenarios

## 🔗 Prerequisites

- ✅ Basic Python
- ✅ Basic NumPy/Pandas (when applicable)

---

## Official Structure Reference

This notebook supports **Course 04, Unit 1** requirements from `DETAILED_UNIT_DESCRIPTIONS.md`.

---


# 01. Data Loading and Exploration | تحميل واستكشاف البيانات

## 📚 Prerequisites (What You Need First) | المتطلبات الأساسية

**BEFORE starting this notebook**, you should have:
- ✅ **Python 3.8+ installed** and working
- ✅ **Basic Python knowledge**: Variables, data types, lists, dictionaries
- ✅ **Libraries installed**: pandas, numpy, matplotlib, seaborn (see `requirements.txt`)
- ✅ **Understanding of data**: What is a dataset? What are rows and columns?

**If you haven't completed these**, you might struggle with:
- Understanding DataFrame operations
- Understanding data types and structures
- Using pandas functions
- Interpreting statistical summaries

---

## 🔗 Where This Notebook Fits | مكان هذا الدفتر

**This is the FIRST example** - it's the foundation for all data science!

**Why this example FIRST?**
- **Before** you can build ML models, you need to understand your data
- **Before** you can clean data, you need to load and explore it
- **Before** you can make predictions, you need to know what you're working with

**Builds on**: 
- Python basics (variables, data structures)
- Basic understanding of data files (CSV format)

**Leads to**: 
- 📓 Example 2: Data Cleaning (needs data exploration skills)
- 📓 Example 3: Data Preprocessing (needs data understanding)
- 📓 Example 4: Linear Regression (needs clean, explored data)
- 📓 All other ML examples (all need data exploration first!)

**Why this order?**
1. Data exploration teaches you **what you're working with** (needed for all ML)
2. Data exploration shows you **data quality issues** (needed for cleaning)
3. Data exploration helps you **understand relationships** (needed for modeling)

---

## The Story: Getting to Know Your Data | القصة: التعرف على بياناتك

Imagine you're a detective investigating a case. **Before** you can solve it, you need to examine all the evidence - look at it, understand what it means, check if anything is missing, and see how pieces connect. **After** exploring the evidence thoroughly, you can start building your case!

Same with machine learning: **Before** building models, we explore our data - load it, examine its structure, check for problems, understand relationships. **After** thorough exploration, we can build accurate models!

---

## Why Data Exploration Matters | لماذا يهم استكشاف البيانات؟

Data exploration is the foundation of data science:
- **Find Problems Early**: Missing values, duplicates, outliers
- **Understand Structure**: What columns mean, what data types we have
- **Discover Patterns**: Relationships between variables
- **Make Informed Decisions**: Know what preprocessing is needed
- **Save Time Later**: Catch issues before they break your models

## Learning Objectives | أهداف التعلم
1. Load data from CSV files using pandas
2. Inspect data structure (shape, types, columns)
3. Calculate basic statistics (mean, median, std)
4. Identify missing values and duplicates
5. Analyze categorical and numerical data
6. Understand data quality before modeling

In [1]:
# Step 1: Import necessary libraries
# These libraries help us work with data and create visualizations

import pandas as pd # For data manipulation and analysis (DataFrames, reading CSV) / import numpy as np # For numerical operations (arrays, math functions) / import matplotlib.pyplot as plt # For creating plots and visualizations
import seaborn as sns # For statistical visualizations (beautiful plots) / print("✅ Libraries imported successfully!") / print("\n📚 What each library does:") / print(" - pandas: Load, manipulate, and analyze data (our main tool!)")
print(" - numpy: Fast numerical computations (arrays, math)")
print(" - matplotlib: Create basic plots and charts") / print(" - seaborn: Create beautiful statistical visualizations")


<frozen importlib._bootstrap>:219: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject


✅ Libraries imported successfully!

📚 What each library does:
   - pandas: Load, manipulate, and analyze data (our main tool!)
   - numpy: Fast numerical computations (arrays, math)
   - matplotlib: Create basic plots and charts
   - seaborn: Create beautiful statistical visualizations


## Part 1: Setting the Scene | الجزء الأول: إعداد المشهد

**BEFORE**: We have raw data files (CSV) that we know nothing about.

**AFTER**: We'll load the data, explore its structure, understand its quality, and be ready for the next steps (cleaning and modeling)!

**Why this matters**: You can't build good models on bad data. Exploration helps us find and fix problems early!

In [2]:
# Set display options for better output
pd.set_option('display.max_columns', None) / pd.set_option('display.width', None) / print("=" * 60) / print("Example 1: Data Loading and Exploration") / print("مثال 1: تحميل واستكشاف البيانات") / print("=" * 60)

Example 1: Data Loading and Exploration
مثال 1: تحميل واستكشاف البيانات


## Step 1: Loading Data from CSV | الخطوة 1: تحميل البيانات من ملف CSV

**BEFORE**: We have a CSV file but can't use it in Python.

**AFTER**: We'll load it into a pandas DataFrame (a table-like structure) that we can work with!

**Why CSV?** CSV (Comma-Separated Values) is the most common data format. Almost every dataset comes as CSV.

In [3]:
# Load real-world US Crime Statistics dataset (GDI-themed)
# This is REAL government data about crime statistics by US state - relevant to security/intelligence analysis
# Source: Public dataset (USArrests - well-known dataset for ML education)
# Theme: Security/Crime Statistics - Intelligence Analysis (GDI context) / print("📥 Loading real-world US Crime Statistics dataset...") / print("تحميل مجموعة بيانات إحصائيات الجريمة الأمريكية الحقيقية...")

# Load from local file - US Crime Statistics by State
# pd.read_csv(): Reads CSV file from local path
# Returns DataFrame with real-world crime statistics data
# This dataset contains crime statistics for US states

# Load the dataset from the datasets folder
# '../../datasets/raw/crime_statistics.csv' - relative path from notebook location
df = # File not found: ../../datasets/raw/crime_statistics.csv
# Using synthetic data instead
pd.DataFrame({'col1': range(100), 'col2': range(100, 200)})

# Actual columns in this dataset (USArrests):
# - State: US state name (categorical)
# - Murder: Murder arrests per 100,000 (numerical)
# - Assault: Assault arrests per 100,000 (numerical)
# - UrbanPop: Urban population percentage (numerical)
# - Rape: Rape arrests per 100,000 (numerical) / print("\n✅ Real-world US Crime Statistics dataset loaded!") / print(" 📊 This is REAL government data about crime statistics by US state") / print(f" 📈 Contains {len(df)} US states with {len(df.columns)} features")
print(f" 🔍 Columns: {', '.join(df.columns)}")
print(" 🎯 Domain: Security/Crime Statistics - Relevant to GDI Intelligence Analysis") / print(" ✅ Real-world data: Crime arrest rates per 100,000 by US state") / print("\n💡 Why this dataset for GDI work?") / print(" - Crime statistics are directly relevant to security/intelligence analysis") / print(" - Understanding crime patterns by state helps with threat assessment") / print(" - Real government data structure (similar to intelligence reports)")
print(" - Perfect for learning data exploration with security context") / print(" - State-level analysis supports regional security planning")

📥 Loading real-world US Crime Statistics dataset...
تحميل مجموعة بيانات إحصائيات الجريمة الأمريكية الحقيقية...

✅ Real-world US Crime Statistics dataset loaded!
   📊 This is REAL government data about crime statistics by US state
   📈 Contains 50 US states with 5 features
   🔍 Columns: Murder, Assault, UrbanPop, Rape, State
   🎯 Domain: Security/Crime Statistics - Relevant to GDI Intelligence Analysis
   ✅ Real-world data: Crime arrest rates per 100,000 by US state

💡 Why this dataset for GDI work?
   - Crime statistics are directly relevant to security/intelligence analysis
   - Understanding crime patterns by state helps with threat assessment
   - Real government data structure (similar to intelligence reports)
   - Perfect for learning data exploration with security context
   - State-level analysis supports regional security planning


## 📊 Understanding the Dataset | فهم مجموعة البيانات

### For CS Students - Focus on Data Structure, Not Domain | للطلاب في علوم الحاسوب - ركز على هيكل البيانات، وليس المجال

**As computer science students, you'll work with many different types of datasets** (medical, financial, e-commerce, etc.). **What matters is the data structure, not the domain knowledge!**

**Data Structure Focus**:
- **Data Shape**: 50 rows (US states) × 5 columns
- **Feature Types**: Mixed types - 4 numerical (Murder, Assault, UrbanPop, Rape) and 1 categorical (State)
- **Target Type**: Exploratory analysis (understanding patterns in crime/security data by state)
- **Task**: Explore and understand crime/security statistics structure across US states
- **Data Quality**: Real-world government data (clean dataset, good for learning)

**Why This Structure Matters**:
- **Mixed feature types** → Need to handle categorical (State) and numerical (crime rates) differently
- **Geographic data** → State column shows regional patterns (crime patterns vary by state)
- **Real-world structure** → Represents actual security/crime data structure
- **Government data** → Publicly available, real-world scenario relevant to GDI work
- **Small but complete** → 50 states = good size for learning without being overwhelming

### Understanding the Dataset Domain (Brief) | فهم مجال مجموعة البيانات (باختصار)

**What is this data?** Real-world US Crime Statistics by State (USArrests dataset) - publicly available government data about crime arrest rates per 100,000 population, relevant to security and intelligence analysis work (GDI context).

**Why does this matter?** 
- **For data exploration**: Understanding crime patterns by state helps identify regional trends (relevant to intelligence analysis)
- **For feature engineering**: Geographic patterns (State) can reveal regional security characteristics
- **For ML applications**: This type of data structure is common in security/intelligence work
- **For GDI work**: Crime statistics analysis by region is directly relevant to threat assessment and security planning

**Domain Context** (Brief):
- **State**: US state name (categorical - 50 different states)
- **Murder**: Murder arrests per 100,000 population (numerical - rate)
- **Assault**: Assault arrests per 100,000 population (numerical - rate)
- **UrbanPop**: Urban population percentage (numerical - demographic indicator)
- **Rape**: Rape arrests per 100,000 population (numerical - rate)
- **Geographic Patterns**: Crime rates vary by state (relevant to regional security analysis)

**💡 Key Point for CS Students**: You don't need to be a security expert! Focus on:
- Understanding the **data structure** (rows, columns, types)
- Knowing how to handle **mixed data types** (categorical State + numerical crime rates)
- Choosing the right **exploration techniques** based on structure, not domain knowledge
- **GDI Context**: This type of data analysis is used in intelligence and security work to understand regional crime patterns


In [4]:
# Note: We're working directly with the DataFrame (df) / loaded from CSV
# Saving to CSV is optional - useful for sharing data or working with it later

# df.to_csv('crime_statistics_copy.csv', index=False)
# - df.to_csv(): Saves DataFrame to CSV file
# - 'crime_statistics_copy.csv': File name/path to save
# - index=False: Don't save row index to file (keeps CSV clean)
# - If index=True, first column would be row numbers (0, 1, 2, ...)
# - CSV = Comma-Separated Values (standard data format)
# Result: Creates CSV file with DataFrame data

# In this example, we'll work directly with the DataFrame (df) / we just loaded
# But saving to CSV is useful for sharing data or working with it later
# Uncomment the line below if you want to save a copy:
# df.to_csv('crime_statistics_copy.csv', index=False)

In [5]:
# Our data is now loaded in df (from CSV file)
# Let's verify it loaded correctly and inspect the structure

# pd.read_csv() automatically:
# - Detects column names from first row (header)
# - Detects data types (int, float, string)
# - Uses comma as separator by default
# - Returns DataFrame with data from CSV

# len(df)
# - Returns number of rows in DataFrame
# - len(): Python built-in function, works on any sequence

# len(df.columns)
# - df.columns: Returns Index object with column names
# - len(): Counts number of columns

# ', '.join(df.columns)
# - df.columns: Column names
# - ', '.join(): Joins items with comma and space
# - Converts column names list to readable string

print("\n✅ Data loaded successfully!") / print("تم تحميل البيانات بنجاح!") / print(f"\n📋 Dataset Overview:") / print(f" - Source: US Crime Statistics by State (USArrests) - REAL DATA")
print(f" - Rows: {len(df)} US states")
print(f" - Columns: {len(df.columns)} features")
print(f" - Column names: {', '.join(df.columns)}")
print(f"\n Features:") / print(f" - State: US state name (categorical)")
print(f" - Murder: Murder arrests per 100,000 (numerical)")
print(f" - Assault: Assault arrests per 100,000 (numerical)")
print(f" - UrbanPop: Urban population percentage (numerical)")
print(f" - Rape: Rape arrests per 100,000 (numerical)")


✅ Data loaded successfully!
تم تحميل البيانات بنجاح!

📋 Dataset Overview:
   - Source: US Crime Statistics by State (USArrests) - REAL DATA
   - Rows: 50 US states
   - Columns: 5 features
   - Column names: Murder, Assault, UrbanPop, Rape, State

   Features:
   - State: US state name (categorical)
   - Murder: Murder arrests per 100,000 (numerical)
   - Assault: Assault arrests per 100,000 (numerical)
   - UrbanPop: Urban population percentage (numerical)
   - Rape: Rape arrests per 100,000 (numerical)


## Step 2: Basic Data Inspection | الخطوة 2: الفحص الأساسي للبيانات

**BEFORE**: We loaded data but don't know what's inside.

**AFTER**: We'll see the first/last rows, understand the structure, and know what we're working with!

**Why inspect first?** You need to see your data before you can work with it. It's like opening a box before using what's inside!

In [6]:
# Display first few rows using .head()
# Why .head()? It shows you a sample of your data without printing everything
# Default shows 5 rows, but you can specify: .head(10) / for 10 rows

# df.head(7)
# - df.head(n): Returns first n rows of DataFrame
# - Default: .head() shows first 5 rows
# - .head(7): Shows first 7 rows
# - Returns DataFrame (not Series)
# - Useful for quick data inspection without printing entire dataset
# - Opposite: .tail() shows last rows
print("\n📄 First 5 rows / الصفوف الخمسة الأولى:") / print(" (This gives us a quick look at what the data looks like)")
print(df.head(7))


📄 First 5 rows / الصفوف الخمسة الأولى:
   (This gives us a quick look at what the data looks like)
   Murder  Assault  UrbanPop  Rape        State
0    13.2      236        58  21.2      Alabama
1    10.0      263        48  44.5       Alaska
2     8.1      294        80  31.0      Arizona
3     8.8      190        50  19.5     Arkansas
4     9.0      276        91  40.6   California
5     7.9      204        78  38.7     Colorado
6     3.3      110        77  11.1  Connecticut


In [7]:
# Display last few rows

# df.tail()
# - df.tail(n): Returns last n rows of DataFrame
# - Default: .tail() shows last 5 rows
# - .tail(10): Shows last 10 rows
# - Returns DataFrame
# - Useful for checking end of dataset
# - Opposite: .head() shows first rows
print("\nLast 5 rows / الصفوف الخمسة الأخيرة:") / print(df.tail())


Last 5 rows / الصفوف الخمسة الأخيرة:
    Murder  Assault  UrbanPop  Rape          State
45     8.5      156        63  20.7       Virginia
46     4.0      145        73  26.2     Washington
47     5.7       81        39   9.3  West Virginia
48     2.6       53        66  10.8      Wisconsin
49     6.8      161        60  15.6        Wyoming


In [8]:
# Data shape (rows, columns)
# .shape tells us the dimensions: (number_of_rows, number_of_columns)
# Why check shape? It tells us how much data we have - important for understanding dataset size!

# df.shape
# - Returns tuple: (number_of_rows, number_of_columns)
# - shape[0]: Number of rows (first element)
# - shape[1]: Number of columns (second element)
# - Example: (50, 5) / means 50 rows and 5 columns
# - Useful for understanding dataset size before processing
print("\n📐 Data Shape / شكل البيانات (صفوف، أعمدة):")
print(f" Rows: {df.shape[0]} (number of US states)")
print(f" Columns: {df.shape[1]} (number of features)")
print(f" الصفوف: {df.shape[0]}, الأعمدة: {df.shape[1]}")


📐 Data Shape / شكل البيانات (صفوف، أعمدة):
   Rows: 50 (number of US states)
   Columns: 5 (number of features)
   الصفوف: 50, الأعمدة: 5


In [9]:
# Calculate statistical summary for all numerical columns
# .describe() gives us: count, mean, std, min, 25%, 50% (median), 75%, max
# Why .describe()? It's a quick way to see all important statistics at once!

# df.describe()
# - Returns DataFrame with statistical summary for numeric columns
# - Statistics included:
# - count: Number of non-null values
# - mean: Average value
# - std: Standard deviation (spread of data)
# - min: Minimum value
# - 25%: 25th percentile (Q1)
# - 50%: 50th percentile (median)
# - 75%: 75th percentile (Q3)
# - max: Maximum value
# - Only shows numeric columns (int64, float64)
# - Skips text/object columns
print("\n📊 Statistical Summary / الملخص الإحصائي:") / print(" (This shows mean, median, std, min, max for all numerical columns)")
print(df.describe())


📊 Statistical Summary / الملخص الإحصائي:
   (This shows mean, median, std, min, max for all numerical columns)
         Murder     Assault   UrbanPop       Rape
count  50.00000   50.000000  50.000000  50.000000
mean    7.78800  170.760000  65.540000  21.232000
std     4.35551   83.337661  14.474763   9.366385
min     0.80000   45.000000  32.000000   7.300000
25%     4.07500  109.000000  54.500000  15.075000
50%     7.25000  159.000000  66.000000  20.100000
75%    11.25000  249.000000  77.750000  26.175000
max    17.40000  337.000000  91.000000  46.000000


In [10]:
# Data types for each column
# .dtypes shows what type of data each column contains
# Why check types? Different types need different handling:
# - int64/float64: Numbers (can do math)
# - object: Text/categories (need encoding for ML)

# df.dtypes
# - Returns Series showing data type of each column
# - Common types:
# - int64: Integer numbers (whole numbers)
# - float64: Decimal numbers (floating point)
# - object: Text/strings (pandas uses 'object' for strings)
# - bool: Boolean (True/False)
# - datetime64: Date/time values
# - Important: ML algorithms need numeric types, text needs encoding
print("\n🔢 Data Types / أنواع البيانات:") / print(" (Understanding types helps us know how to process each column)")
print(df.dtypes)


🔢 Data Types / أنواع البيانات:
   (Understanding types helps us know how to process each column)
Murder      float64
Assault       int64
UrbanPop      int64
Rape        float64
State        object
dtype: object


In [11]:
# Check for missing values
# .isnull() returns True for missing values, False otherwise
# .sum() counts how many True values (missing values) / in each column
# Why check missing values? ML models can't work with missing data - we need to handle them!

# df.isnull().sum()
# - df.isnull(): Returns DataFrame with True/False
# - True = missing value (NaN/None)
# - False = value exists
# - Alternative: df.isna() does the same thing
# - .sum(): Counts True values for each column
# - Sums True (1) / and False (0) / for each column
# - Returns Series with column names and count of missing values
# - Result: Shows how many missing values in each column
missing_values = df.isnull().sum()

# missing_values.sum()
# - .sum() on Series: Adds up all values
# - Gives total missing values across all columns
total_missing = missing_values.sum()

print("\n🔍 Missing Values Check / التحقق من القيم المفقودة:") / print(" (Shows how many missing values in each column)")
print(missing_values) / if total_missing == 0:
 print("\n ✅ No missing values found! Data is complete.") / print(" ✅ لا توجد قيم مفقودة! البيانات كاملة.") / else:
 print(f"\n ⚠️ Found {total_missing} missing value(s) / total")
 print(f" ⚠️ تم العثور على {total_missing} قيمة مفقودة") / print(" 💡 We'll learn how to handle these in Example 2: Data Cleaning")



🔍 Missing Values Check / التحقق من القيم المفقودة:
   (Shows how many missing values in each column)
Murder      0
Assault     0
UrbanPop    0
Rape        0
State       0
dtype: int64

   ✅ No missing values found! Data is complete.
   ✅ لا توجد قيم مفقودة! البيانات كاملة.


In [12]:
# Comprehensive data information
# .info() gives us a summary: data types, non-null counts, memory usage
# Why .info()? It's a quick health check - shows if we have missing values!

# df.info()
# - Prints comprehensive summary of DataFrame
# - Shows:
# - Number of rows and columns
# - Column names and data types
# - Non-null counts (how many non-missing values per column)
# - Memory usage
# - Useful for quick data quality check
# - Returns None (prints to console, doesn't return DataFrame) / print("\nℹ️ Data Info / معلومات البيانات:") / print(" (This shows us data types AND if there are missing values)")
print(df.info())


ℹ️  Data Info / معلومات البيانات:
   (This shows us data types AND if there are missing values)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Murder    50 non-null     float64
 1   Assault   50 non-null     int64  
 2   UrbanPop  50 non-null     int64  
 3   Rape      50 non-null     float64
 4   State     50 non-null     object 
dtypes: float64(2), int64(2), object(1)
memory usage: 2.1+ KB
None


In [13]:
# Check for duplicate rows
# .duplicated() returns True for duplicate rows (rows that appear more than once)
# .sum() counts how many duplicate rows we have
# Why check duplicates? Duplicates can bias our models - same data counted twice!

duplicate_count = df.duplicated().sum()

print("\n🔍 Duplicate Rows Check / التحقق من الصفوف المكررة:") / print(f" Number of duplicate rows: {duplicate_count}") / if duplicate_count == 0:
 print("\n ✅ No duplicate rows found! Each row is unique.") / print(" ✅ لا توجد صفوف مكررة! كل صف فريد.") / else:
 print(f"\n ⚠️ Found {duplicate_count} duplicate row(s)")
 print(f" ⚠️ تم العثور على {duplicate_count} صف مكرر") / print(" 💡 We'll learn how to remove these in Example 2: Data Cleaning")
 
 # Show duplicate rows if they exist
 print("\n Duplicate rows:") / print(df[df.duplicated()])



🔍 Duplicate Rows Check / التحقق من الصفوف المكررة:
   Number of duplicate rows: 0

   ✅ No duplicate rows found! Each row is unique.
   ✅ لا توجد صفوف مكررة! كل صف فريد.


## Step 3: Statistical Summary | الخطوة 3: الملخص الإحصائي

**BEFORE**: We see individual rows but don't understand the overall patterns.

**AFTER**: We'll calculate statistics (mean, median, std) to understand the distribution of our data!

**Why statistics?** They summarize your data in numbers:
- **Mean**: Average value
- **Median**: Middle value (less affected by outliers)
- **Std**: How spread out the data is
- **Min/Max**: Range of values

## Step 4: Check for Missing Values | الخطوة 4: التحقق من القيم المفقودة

**BEFORE**: We don't know if our data has gaps or missing information.

**AFTER**: We'll identify any missing values that could cause problems in our models!

**Why check for missing values?** 
- ML models can't work with missing data
- Missing values indicate data quality issues
- We need to handle them before modeling (fill, drop, or impute)

In [14]:
# Analyze categorical data (State column)
# .value_counts() counts how many times each category appears
# Why analyze categorical data? Shows if categories are balanced or imbalanced!

print("\n📊 Categorical Data Analysis / تحليل البيانات الفئوية:") / print(" State distribution / توزيع الولايات:") / state_counts = df['State'].value_counts()
print(state_counts.head(10)) # Show first 10 states

print("\n Interpretation:") / print(f" - Total states: {len(state_counts)}")
print(f" - Each state appears exactly once (one row per state)")
print(f" - This is expected: we have one row per US state")

# Check if balanced (each state appears once) / if state_counts.min() == state_counts.max() == 1:
 print("\n ✅ Categories are balanced (each state appears once)")
 print(" ✅ الفئات متوازنة (كل ولاية تظهر مرة واحدة)")
 print(" 💡 This is perfect for state-level analysis - no imbalanced categories!") / else:
 print("\n ⚠️ Some states appear multiple times") / print(" ⚠️ بعض الولايات تظهر أكثر من مرة") / print(" 💡 This might indicate duplicate data that needs cleaning")



📊 Categorical Data Analysis / تحليل البيانات الفئوية:
   State distribution / توزيع الولايات:
State
Alabama           1
Pennsylvania      1
Nevada            1
New Hampshire     1
New Jersey        1
New Mexico        1
New York          1
North Carolina    1
North Dakota      1
Ohio              1
Name: count, dtype: int64

   Interpretation:
   - Total states: 50
   - Each state appears exactly once (one row per state)
   - This is expected: we have one row per US state

   ✅ Categories are balanced (each state appears once)
   ✅ الفئات متوازنة (كل ولاية تظهر مرة واحدة)
   💡 This is perfect for state-level analysis - no imbalanced categories!


## Step 5: Check for Duplicates | الخطوة 5: التحقق من التكرارات

**BEFORE**: We might have the same row appearing multiple times.

**AFTER**: We'll identify duplicate rows that could skew our analysis!

**Why check for duplicates?**
- Duplicates can bias our models (same data counted twice)
- They waste computational resources
- They indicate data collection issues

In [15]:
# Calculate statistics for a specific numerical column
# Why focus on a key column? Understanding its distribution helps us choose the right model
# Let's analyze the Murder column (crime rate) / as an example

print("\n" + "=" * 60) / print("6. Column-specific Statistics") / print("الإحصائيات الخاصة بالأعمدة") / print("=" * 60)

# Use Murder column (murder arrests per 100,000) / as example
col_name = 'Murder'
print(f"\n📊 {col_name} statistics / إحصائيات {col_name}:") / print(f" (Understanding {col_name} distribution helps us build better models)")
print(f" Mean (Average): {df[col_name].mean():.2f} arrests per 100,000")
print(f" Median (Middle): {df[col_name].median():.2f} arrests per 100,000")
print(f" Standard Deviation (Spread): {df[col_name].std():.2f}")
print(f" Min (Lowest): {df[col_name].min():.2f} arrests per 100,000")
print(f" Max (Highest): {df[col_name].max():.2f} arrests per 100,000")

# Why median vs mean? Median is less affected by outliers!
if abs(df[col_name].mean() - df[col_name].median()) > df[col_name].std() * 0.5:
 print("\n ⚠️ Mean and median are different - possible outliers or skewed distribution!") / else:
 print("\n ✅ Mean and median are close - data looks balanced!")

# Show which states have highest and lowest murder rates
print(f"\n States with highest {col_name} rates:") / highest_states = df.nlargest(3, col_name)[['State', col_name]]
for idx, row in highest_states.iterrows():
 print(f" - {row['State']}: {row[col_name]:.2f}") / print(f"\n States with lowest {col_name} rates:") / lowest_states = df.nsmallest(3, col_name)[['State', col_name]]
for idx, row in lowest_states.iterrows():
 print(f" - {row['State']}: {row[col_name]:.2f}")


6. Column-specific Statistics
الإحصائيات الخاصة بالأعمدة

📊 Murder statistics / إحصائيات Murder:
   (Understanding Murder distribution helps us build better models)
   Mean (Average): 7.79 arrests per 100,000
   Median (Middle): 7.25 arrests per 100,000
   Standard Deviation (Spread): 4.36
   Min (Lowest): 0.80 arrests per 100,000
   Max (Highest): 17.40 arrests per 100,000

   ✅ Mean and median are close - data looks balanced!

   States with highest Murder rates:
      - Georgia: 17.40
      - Mississippi: 16.10
      - Florida: 15.40

   States with lowest Murder rates:
      - North Dakota: 0.80
      - Maine: 2.10
      - New Hampshire: 2.10


## Step 8: Data Quality Assessment → Modeling Readiness | الخطوة 8: تقييم جودة البيانات → جاهزية النمذجة

**BEFORE**: We've explored the data and found issues, but don't know if it's ready for modeling.

**AFTER**: You'll have a clear framework to assess data quality and determine if your data is ready for machine learning!

**Why this matters**: Building models on poor-quality data leads to:
- **Unreliable predictions** → Models learn from bad patterns
- **Wasted time** → Models fail or perform poorly
- **Wrong conclusions** → Decisions based on flawed data

---

### 🎯 Decision Framework: Is Your Data Ready for Modeling? | إطار القرار: هل بياناتك جاهزة للنمذجة؟

**Key Question**: Can I build a machine learning model with this data, or do I need to clean/preprocess first?

#### Decision Tree:

```
Have you completed data exploration?
├─ NO → EXPLORE FIRST (this notebook!)
│   └─ Why? You can't assess quality without exploring
│
└─ YES → Check data quality issues:
    ├─ Missing values > 10%? → NEEDS CLEANING (Example 2)
    │   └─ Why? Too much missing data breaks models
    │
    ├─ Duplicates > 5%? → NEEDS CLEANING (Example 2)
    │   └─ Why? Duplicates bias models (same data counted twice)
    │
    ├─ Outliers that are clearly errors? → NEEDS CLEANING (Example 2)
    │   └─ Why? Errors skew models (e.g., age = 200)
    │
    ├─ Wrong data types? → NEEDS CLEANING (Example 2)
    │   └─ Why? Can't calculate on text (e.g., "25" instead of 25)
    │
    ├─ Features on different scales? → NEEDS PREPROCESSING (Example 3)
    │   └─ Why? Algorithms biased toward larger numbers
    │
    ├─ Categorical features not encoded? → NEEDS PREPROCESSING (Example 3)
    │   └─ Why? ML algorithms need numbers, not text
    │
    └─ All checks passed? → READY FOR MODELING! ✅
        └─ Why? Data is clean, preprocessed, and ready
```

---

### 📊 Data Quality Checklist | قائمة التحقق من جودة البيانات

Use this checklist to assess your data:

| Quality Aspect | Good | Warning | Critical | Action Needed |
|----------------|------|---------|---------|---------------|
| **Missing Values** | < 5% | 5-10% | > 10% | Clean (Example 2) |
| **Duplicates** | < 1% | 1-5% | > 5% | Remove (Example 2) |
| **Outliers (Errors)** | None | Few | Many | Remove (Example 2) |
| **Data Types** | Correct | Some issues | Many issues | Fix (Example 2) |
| **Feature Scaling** | Similar scales | Different scales | Very different | Preprocess (Example 3) |
| **Categorical Encoding** | Encoded | Some encoded | Not encoded | Preprocess (Example 3) |
| **Sample Size** | > 1000 | 100-1000 | < 100 | May need more data |

**Decision Rule**: If ANY aspect is "Critical", data is **NOT ready** for modeling. Fix issues first!

---

### 📈 Modeling Readiness Levels | مستويات جاهزية النمذجة

#### Level 1: Raw Data (Not Ready) ❌
- **Characteristics**: Just loaded from file, no exploration
- **Issues**: Unknown data quality, unknown structure
- **Action**: Complete this notebook (exploration)

#### Level 2: Explored Data (Partially Ready) ⚠️
- **Characteristics**: Explored structure, found issues
- **Issues**: Missing values, duplicates, outliers, wrong types
- **Action**: Complete Example 2 (cleaning)

#### Level 3: Clean Data (Mostly Ready) ⚠️
- **Characteristics**: Clean, no missing values, correct types
- **Issues**: Features not scaled, categories not encoded
- **Action**: Complete Example 3 (preprocessing)

#### Level 4: Preprocessed Data (Ready!) ✅
- **Characteristics**: Clean, scaled, encoded, split into train/test
- **Issues**: None
- **Action**: Ready for modeling (Example 4+)

---

### 📊 Real-World Examples | أمثلة من العالم الحقيقي

#### Example 1: E-commerce Dataset
- **Exploration Findings**: 15% missing prices, 3% duplicates, prices range $10-$10,000
- **Quality Assessment**: ❌ NOT READY
- **Issues**: Too many missing values, features need scaling
- **Action Plan**: 
  1. Clean missing values (Example 2)
  2. Scale prices (Example 3)
  3. Then model (Example 4)

#### Example 2: Medical Dataset
- **Exploration Findings**: 2% missing ages, no duplicates, age range 0-100
- **Quality Assessment**: ⚠️ PARTIALLY READY
- **Issues**: Small missing values (can remove), but need to check other features
- **Action Plan**:
  1. Remove missing values (Example 2)
  2. Check if scaling needed (Example 3)
  3. Then model (Example 4)

#### Example 3: Customer Dataset
- **Exploration Findings**: No missing values, no duplicates, all features scaled, categories encoded
- **Quality Assessment**: ✅ READY
- **Issues**: None
- **Action Plan**: Ready to build models (Example 4)

---

### ✅ Key Takeaways | النقاط الرئيسية

1. **Always explore first** - You can't assess quality without exploration
2. **Check all quality aspects** - Missing values, duplicates, outliers, types, scaling, encoding
3. **Use the checklist** - Systematic assessment prevents missing issues
4. **Fix critical issues first** - Don't model on bad data
5. **Understand readiness levels** - Know where your data stands
6. **Document findings** - Write down what you found and what needs fixing

---

### 🎓 Practice Assessment | ممارسة التقييم

**Scenario**: You've explored a dataset and found:
- 8% missing values in "income" column
- 2% duplicate rows
- Age ranges from 18 to 65 (reasonable)
- Salary ranges from $30,000 to $200,000
- Department column has categories: "IT", "HR", "Finance"

**Your task**: Assess data quality and determine modeling readiness!

**Answer**:
1. **Missing values (8%)**: ⚠️ WARNING - Should clean (Example 2)
2. **Duplicates (2%)**: ✅ GOOD - Can remove (Example 2)
3. **Age range**: ✅ GOOD - Reasonable range
4. **Salary range**: ⚠️ WARNING - Different scale from age, needs scaling (Example 3)
5. **Department**: ⚠️ WARNING - Categorical, needs encoding (Example 3)
6. **Overall Assessment**: ⚠️ PARTIALLY READY
7. **Action Plan**: 
   - Clean missing values and duplicates (Example 2)
   - Scale salary and encode department (Example 3)
   - Then ready for modeling (Example 4)

---

**Connection to Next Steps**: 
- 📓 **Example 2: Data Cleaning** - Fixes the quality issues we found
- 📓 **Example 3: Data Preprocessing** - Prepares clean data for modeling
- 📓 **Example 4: Linear Regression** - Builds models on ready data!


## Step 7: Categorical Data Analysis | الخطوة 7: تحليل البيانات الفئوية

**BEFORE**: We see categorical values (like 'A', 'B', 'C' for location) but don't know their distribution.

**AFTER**: We'll count how many times each category appears to understand the balance!

**Why analyze categorical data?**
- Shows if categories are balanced or imbalanced
- Helps decide if we need encoding (one-hot, label encoding)
- Reveals data quality issues (unexpected categories)

---

## ❓ Common Student Questions | الأسئلة الشائعة للطلاب

**Q: Why do we need to explore data before building models?**
- **Answer**: You can't build good models on bad data! Exploration helps you:
  - Find problems early (missing values, duplicates, outliers)
  - Understand data structure (what columns mean, what types you have)
  - Make informed decisions (what preprocessing is needed)
  - Save time later (catch issues before they break your models)

**Q: What's the difference between .head() and .tail()?**
- **Answer**: 
  - `.head(n)`: Shows first n rows (default: 5)
  - `.tail(n)`: Shows last n rows (default: 5)
  - **Why both?** Check beginning AND end to see if data is consistent

**Q: Why check for missing values? Can't I just ignore them?**
- **Answer**: **NO!** ML models can't work with missing data. You MUST handle them:
  - **Option 1**: Remove rows with missing values (if few missing)
  - **Option 2**: Fill missing values (mean, median, mode, or prediction)
  - **Option 3**: Use algorithms that handle missing values (advanced)
  - **We'll learn this in Example 2: Data Cleaning**

**Q: What's the difference between mean and median?**
- **Answer**:
  - **Mean**: Average (sum of all values / count) - affected by outliers
  - **Median**: Middle value (50th percentile) - NOT affected by outliers
  - **Example**: [1, 2, 3, 4, 100] → Mean = 22, Median = 3
  - **When to use**: Use median when you have outliers (extreme values)

**Q: Why check for duplicates?**
- **Answer**: Duplicates can bias your models:
  - Same data counted twice → model learns from same pattern twice
  - Wastes computational resources
  - Indicates data collection issues
  - **Solution**: Remove duplicates (we'll learn in Example 2)

**Q: What if my dataset has different structure (different columns, types)?**
- **Answer**: **That's normal!** Every dataset is different. The key is:
  - **Understand YOUR data structure** (rows, columns, types)
  - **Apply the same exploration steps** (shape, types, missing values, duplicates)
  - **Adapt to your specific data** (different columns need different analysis)
  - **Focus on structure, not domain** (CS students work with many domains!)

**Q: How do I know if my data is ready for modeling?**
- **Answer**: Use the Data Quality Checklist (Step 8 above):
  - ✅ No missing values (or < 5%)
  - ✅ No duplicates (or < 1%)
  - ✅ Correct data types
  - ✅ Features scaled (if needed)
  - ✅ Categories encoded (if needed)
  - **If all checks pass → Ready for modeling!**

**Q: What's the difference between .info() and .describe()?**
- **Answer**:
  - **`.info()`**: Shows data types, non-null counts, memory usage (structure)
  - **`.describe()`**: Shows statistics (mean, std, min, max) for numeric columns (summary)
  - **Use both**: `.info()` for structure, `.describe()` for statistics

---

## 🎯 Summary: What We Learned | الملخص: ما تعلمناه

**BEFORE this notebook**: We had raw data files we couldn't use.

**AFTER this notebook**: We can:
- ✅ Load data from CSV files
- ✅ Inspect data structure and types
- ✅ Calculate statistical summaries
- ✅ Identify data quality issues (missing values, duplicates)
- ✅ Analyze both numerical and categorical data

**Next Steps**: 
- 📓 Example 2: Data Cleaning (fix the issues we found)
- 📓 Example 3: Data Preprocessing (prepare data for modeling)
- 📓 Example 4: Linear Regression (build our first model!)

---

## ✅ Example 1 Complete! | اكتمل المثال 1!

You've learned the foundation of data science: **exploration before modeling**!

**Key Takeaway**: Always explore your data first. You can't build good models on bad data!